src/nnet3bin/nnet3-train.cc

```diff
@@ -74,10 +79,50 @@ int main(int argc, char *argv[]) {
     {
       NnetTrainer trainer(train_config, &nnet);
 
+      // extract recurrent output names and their offsets from nnet 
+      std::vector<std::string> recurrent_output_names;
+      std::vector<std::string> recurrent_node_names; 
+      std::vector<int32> recurrent_offsets;
+      GetRecurrentOutputNodeNames(nnet, &recurrent_output_names,
+                                  &recurrent_node_names);
+      GetRecurrentNodeOffsets(nnet, recurrent_node_names, &recurrent_offsets);
+      for (int32 i = 0; i < recurrent_offsets.size(); i++)
+        KALDI_LOG << "recurrent node: " << recurrent_node_names[i] 
+                  << ", offset: " << recurrent_offsets[i];
+
+      // pass on the above info to trainer
```

Comment:
https://google.github.io/styleguide/cppguide.html#Function_Comments

Every function declaration should have comments immediately preceding it that describe what the function does and how to use it. These comments should be descriptive ("Opens the file") rather than imperative ("Open the file"); the comment describes the function, it does not tell the function what to do. In general, these comments do not describe how the function performs its task. Instead, that should be left to comments in the function definition.